yamlfile.c

```diff
@@ -305,29 +159,116 @@ StatusCode fiftyoneDegreesYamlFileIterateWithLimit(
 	int limit,
 	void* state,
 	void(*callback)(KeyValuePair*, uint16_t, void*)) {
+	char* current;
 	FILE *handle;
 	int count = 0;
-
 	StatusCode status = FileOpen(fileName, &handle);
-	FileState fileState = { handle, buffer, 0, 0, length, 0, 0 };
-	if (status == SUCCESS) {
-		// Read the first block
-		status = readNextBlock(&fileState);
-		if (status == SUCCESS ||
-			status == FILE_END_OF_FILE) {
-			while ((count < limit || limit < 0)
-				&& (status == SUCCESS || status == FILE_END_OF_DOCUMENT)) {
-				status = readNextDocument(
-					&fileState,
-					keyValuePairs,
-					collectionSize,
-					state,
-					callback);
-				count++;
+	if (status != SUCCESS) {
+		return status;
+	}
+
+	FileState fileState = { 
+		handle, 
+		buffer, 
+		length,
+		// Set the end and current to 0 to force the next block to be read.
+		0,
+		0,
+		0, 
+		0,
+		0,
+		0,
+		0,
+		false };
+
+	PairState pairState = {
+		keyValuePairs,
+		collectionSize,
+		0,
+		keyValuePairs[0].key,
+		keyValuePairs[0].key + keyValuePairs[0].keyLength - 1 };
+
+	// If there is no limit then set the limit to the largest value to 
+	// avoid checking for negative values in the loop.
+	if (limit < 0) {
+		limit = INT_MAX;
+	}
+
+	while (*(current = readNext(&fileState)) != '\0') {
```

Comment:
`readNext` might return null pointer, so dereferencing `current` is unsafe.

Please see https://github.com/51Degrees/common-cxx/pull/41 for added test cases and a patch.

Reply:
@drasmart please could you modify the branch to address this?